### DISCUSSION TITLE
Problem with timestamp_tcp_setup and timestamp_start in the server_conn of flow

### DISCUSSION THREAD

**Discussion by shane-zhang** [mitmproxy/mitmproxy#5884] [Q&A] [answered]
Hi, we are using mitmproxy (mitmdump and an inline python script) to catch the content and RTT from the proxy to the servers while browsing a specific website.

But we found the RTT resolved by `timestamp_tcp_setup` and `timestamp_start` are not precise compared to the result from tcpdump.

The RTT resolved by `timestamp_tcp_setup` and `timestamp_start` is around 10-30 ms larger than the result calculated from tcpdump.

Is this a limitation imposed by the python programing language?

**Comment by mhils** [comment, marked as answer]
You can see how they are set here: https://github.com/mitmproxy/mitmproxy/blob/4ebccfa2365a15b88d9342641f35d66b53583548/mitmproxy/proxy/server.py#L205-L233

Python can definitely a bit of a limiting factor here under heavy load, but maybe more importantly the time between those two timestamps may also include DNS lookups.

**Reply by shane-zhang** [reply]
Thanks for your answer, that really makes lots of sense. Is there any way that we can rule out the DNS time? Can we tap into the DNS flow and get the DNS start and end time?

**Reply by mhils** [reply]
Currently no, because we don't do DNS resolution ourselves for upstream connections but leave it to the OS.

**Comment by shane-zhang** [comment]
Then the [documentation here](https://docs.mitmproxy.org/stable/api/mitmproxy/connection.html#Server.timestamp_start) is not precise and misleading.

**Reply by mhils** [reply]
Thanks for the heads-up, fixed in https://github.com/mitmproxy/mitmproxy/commit/83c57dc53e55e62fa2f6b4003e73bd664e12ff3d!
